bin/varnishd/cache/cache_conn_pool.c

```diff
@@ -249,17 +254,39 @@ VCP_Rel(struct conn_pool **cpp)
 		(void)shutdown(pfd->fd, SHUT_RDWR);
 		cp->n_kill++;
 	}
-	while (cp->n_kill) {
-		Lck_Unlock(&cp->mtx);
-		(void)usleep(20000);
+
+	Lck_Lock(&conn_pools_mtx);
+	VRBT_INSERT(vrb, &dead_pools, cp);
+	Lck_Unlock(&conn_pools_mtx);
+	Lck_Unlock(&cp->mtx);
+}
+
+void
+VCP_RelPoll(void)
+{
+	struct conn_pool *cp, *cp2;
+
+	ASSERT_CLI();
+	Lck_Lock(&conn_pools_mtx);
+	VRBT_FOREACH_SAFE(cp, vrb, &dead_pools, cp2) {
+		VRBT_REMOVE(vrb, &dead_pools, cp);
+		Lck_Unlock(&conn_pools_mtx);
+		CHECK_OBJ_NOTNULL(cp, CONN_POOL_MAGIC);
+
 		Lck_Lock(&cp->mtx);
+		if (cp->n_kill > 0) {
+			Lck_Unlock(&cp->mtx);
+			continue;
```

Comment:
You can not do this.
You end up removing the connection pool from the tree, but if the kill count is positive we do not add it back.
(I am not saying the correct thing is to add it back, but the problem is that you end up leaking connection pools which might also have other impact such as races).

Reply:
right, my bad, I guess I was not awoken enough when I made this change ^^